src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.State.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    public partial class UseConcreteTypeAnalyzer
+    {
+        private sealed class State
+        {
+            public readonly ConcurrentDictionary<IFieldSymbol, bool> VirtualDispatchFields = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, bool> VirtualDispatchLocals = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, bool> VirtualDispatchParameters = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, bool> MethodsAssignedToDelegate = new(SymbolEqualityComparer.Default);
+
+            public readonly ConcurrentDictionary<IFieldSymbol, ConcurrentBag<ITypeSymbol>> FieldAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, ConcurrentBag<ITypeSymbol>> LocalAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, ConcurrentBag<ITypeSymbol>> ParameterAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, ConcurrentBag<ITypeSymbol>> MethodReturns = new(SymbolEqualityComparer.Default);
+
+            public readonly INamedTypeSymbol Void;
+
+            public State(Compilation compilation)
```

Comment:
I don't think we should store this state across the whole compilation, this will cause memory pressure as we will keep growing state during analysis but don't access most of these entries for the entire duration of compilation.

The state should instead be instantiated in the SymbolStartAction and should use PooledDictionary instead of ConcurrentDictionary fields. It should have a Dispose method that returns all the pooled dictionaries back to the pool, and Dispose can be invoked on the state at the end of SymbolEndAction. I think we can also reduce this from 8 dictionaries to just 2 by having `ISymbol` as the key and use `ISymbol.Kind` property to differentiate between the 4 kinds of member symbols. Given that the dictionaries can only ever have entries for each member of a type, we wouldn't expect it to grow to large.

Reply:
If you want to avoid using pooled dictionaries/hashsets for the fields and continue using concurrent data structures like in the current implementation, one other alternative is to have a dedicated `ObjectPool<State>` created for the compilation, such that you call `State.Free()` in the SymbolEndAction that clears all the dictionaries and subsequently the state instance is returned back to the pool to be used by next SymbolStartAction. I'll let you choose your preferred implementation.

Reply:
Done.